draft-duke-quic-load-balancers.md

```diff
@@ -505,6 +518,208 @@ reserved for config rotation {{config-rotation}}.
+Initial packet. This is in cleartext to be readable for the server, but
+authenticated later in the token.
+
+Original Destination Connection ID: This also in cleartext and authenticated
+later.
+
+Opaque Data: This data MUST contain encrypted information that allows the retry
+service to validate the client's IP address, in accordance with the QUIC
+specification. It MUST also encode a secure hash of the original destination
+connection ID field to verify that this field has not been edited.
+
+Upon receipt of an Initial packet with a token that begins with '0', the retry
+service MUST validate the token in accordance with the QUIC specification. It
+must also verify that the secure hash of the Connect ID is correct. If
+incorrect, the token is invalid.
+
+In active mode, the service MUST issue Retry packets for all Client initial
+packets that contain no token, or a token that has the first bit set to '1'. It
+MUST NOT forward the packet to the server. The service MUST validate all tokens
+with the first bit set to '0'. If successful, the service MUST forward the
+packet with the token intact. If unsuccessful, it MUST drop the packet.
+
+Note that this scheme has a performance drawback. When the retry service is in
+active mode, clients with a token from a NEW_TOKEN frame will suffer a 1-RTT
+penalty even though it has proof of address with its token.
+
+In inactive mode, the service MUST forward all packets that have no token or a
+token with the first bit set to '1'. It MUST validate all tokens with the first
+bit set to '0'. If successful, the service MUST forward the packet with the
+token intact. If unsuccessful, it MUST either drop the packet or forward it
+with the token removed. The latter requires decryption and re-encryption of the
+entire Initial packet to avoid authentication failure. Forwarding the packet
+causes the server to respond without the original_connection_id transport
+parameter, which preserves the normal QUIC signal to the client that there is
+an unauthorized man in the middle.
+
+### Server Requirements
+
+A server behind a non-shared-state retry service MUST NOT send Retry packets.
+
+Tokens sent in NEW_TOKEN frames MUST have the first bit be set to '1'.
+
+If a server receives an Initial Packet with the first bit set to '1', it could
+be from a server-generated NEW_TOKEN frame and should be processed in accordance
+with the QUIC specification. If a server receives an Initial Packet with the
+first bit to '0', it is a Retry token and the server MUST NOT attempt to
+validate it. Instead, it MUST assume the address is validated and MUST extract
+the Original Destination Connection ID, assuming the format described in
+{{nss-service-requirements}}.
+
+## Shared-State Retry Service
+
+A shared-state retry service uses a shared key, so that the server can decode
+the service's retry tokens. It does not require that all traffic pass through
+the Retry service, so servers MAY send Retry packets.
```

Comment:
so servers MAY send Retry packets in response to Initial packets that don't already have a Retry token.

Reply:
Any valid token, not just Retry